socless/integrations.py

```diff
@@ -139,6 +132,51 @@ def apply_conversion_from(self, data, conversion):
         """
         if conversion == "json":
             return json.loads(data)
+        else:
+            raise NotImplementedError("Conversion only supports 'json'")
+
+
+def resolve_template(template_string: str, root_object: dict) -> Any:
+    template = jinja_env.from_string(template_string)
+
+    rendered = template.render(context=root_object)
+    if isinstance(rendered, str):
+        rendered = rendered.replace("&#34;", '"').replace("&#39;", "'")
+    return rendered
+
+
+def convert_legacy_reference_to_template(reference_path: str):
+    """Allow backwards compatibility for legacy socless parameter reference to jinja templates.
+    `resolve_template` jinja translation is supported by this function that converts
+    legacy references to valid jinja templates according to the table below:
+
+        Legacy Reference   | Converted Jinja template
+        ----------------   | ------------------------
+        “$.<ref_path>”     |  “{context.<ref_path>}”
+        “vault:<vault-id>” |  “vault-id | fromvault}”
+        "<something>!json" |  "{<something> | fromjson}"
+    """
+    try:
+        template = reference_path
+
+        if template.startswith(PATH_TOKEN):
+            reference, _, conversion = template.partition(CONVERSION_TOKEN)
+            template = f"context{template[1:]}"
+        elif template.startswith(VAULT_TOKEN):
+            reference, _, conversion = template.partition(CONVERSION_TOKEN)
+            _, _, file_id = reference.partition(VAULT_TOKEN)
+            template = f"vault('{file_id}')"
+        else:
+            return template
```

Comment:
#General comment

Line 163 & 166 repeat the same logic. Not a problem for the code, but could be problem for a future developer if they decide to change that logic but don't realize it needs to change in both places. 

One way to address this would be to invert the flow as such
```
if not template.startswith((PATH_TOKEN, VAULT_TOKEN)):
    return template

reference, _, conversion = template.partition(CONVERSION_TOKEN) # This only happens here now
if template.startswith(VAULT_TOKEN):
   # do vaulty things
elif template.startswith(PATH_TOKEN):
   # do pathy things

# do other things
 
return "what we actually want"
```

This also gives us the benefit of an "early exit" which:
- Helps a code reader know upfront what types of input the function doesn't process. Some folks call this programming style the "bouncer pattern". I find it makes functions easier to read (especially if the function does a lot) but that's mostly a style/preference thing
- Keeps your function from spending cycles running if/else evaluations on an input that its ultimately not going to process. This could be useful if the "unprocessable use-case" is what you get most of the time, which will be true for us as we start to deprecate `vault:` and `$.`. I don't expect much of a performance boost from the early exit in this function though.

Reply:
the duplicated line is an optimization to reduce unnecessary `if` statements as seen in [this PR](https://github.com/twilio-labs/socless_python/pull/32/files#diff-7be42f1c9666464083fbd6f6bbf439de0e9e354962db76f8605dc2cb57c1c494L102-R110)

by inverting the flow, you add two .startswith checks to the beginning and then do those startswiths again when they are present. at a minimum you have two startswiths, at a max you have 4.

without inverting the flow you have a constant 2 startswith checks. 

in my opinion the single duplicated line is so close to the other line that the risk of it being misused is low, especially considering that this entire function is slated for deprecation. 

The readability issue is definitely a personal preference thing because imo seeing
if case
elif case
else

with less than 3 lines of code in each case is pretty readable especially with our checks being startswith CONSTANT

Reply:
Let's explore the optimization conversation a bit more because its an interesting one, especially because of how this `convert_legacy_reference_to_template` function will behave over time. 

It's always good to think about how we can make our code more performant and I 💯 like that you always have this on your mind. However, optimization isn't always as straightforward as structuring the code in the way that seems most optimal. We also have to factor in what the "average input" to our system looks like and how structure responds to it.

Currently, since we're still heavily reliant on `$.` and `vault:` references in our playbooks, the implementation in the PR is actually the most performant because `if template.startswith(PATH_TOKEN)` is our average case input, followed by VAULT_TOKEN, then finally neither. 

However, as we start to migrate away from `$.` / `vault:`,  to the jinja syntax of `{context.}`, the implementation in the PR will start to penalize us because it guarantees that our average case is always handled last. Overtime, any optimizations the PR code hoped to squeeze out through code duplication will be lost post-transition. And while we might deprecate the `$.` and `vault:` features very quickly, the code that handles backward compatibility with those features will likely have to hang around for a while (possibly 6 months to a year) for us to establish full confidence that we can nuke it without braking ours or other peoples' use-cases. There's even a good chance we'll just forget to remove this code altogether as we get busy doing other things.

So overall, optimizations in this particular block of code are tricky. But that really shouldn't bother us because in the larger context of this effort, the code in this function isn't where our big performance bottlenecks are. As long as we aren't doing anything here that is **significantly** slow, we'll likely be fine (admittedly, I haven't timed the performance of my suggestions to validate that it isn't significantly slow. If you decide to time it and it is, please let me know). 

The big takeaway of this post is this: We shouldn't let optimization be the top principle that drives our coding decisions. We should weigh other things as well such as DRY, maintainability, etc.

Reply:
I don't see how the PR code will penalize us in the future, because as long as we have to support `$.` and `vault:` references at the beginning of a string we still have to do 2 .startswith() checks on every string parameter, regardless if its run through jinja before or after (jinja can reference a parameter that uses a vault file syntax). 

This way:
```python
if not template.startswith((PATH_TOKEN, VAULT_TOKEN)):  # step 1 and 2
    return # step 3
```
will return a non-path non-vault string with the same amount of operations as:
```python
        if template.startswith(PATH_TOKEN): # step 1
              _, _, conversion = template.partition(CONVERSION_TOKEN) # not run
              template = f"context{template[1:]}" # not run
              return add_brackets_and_conditionally_add_fromjson( # not run
                  template, bool(conversion) # not run
              ) # not run
        elif template.startswith(VAULT_TOKEN): # step 2
              template = convert_deprecated_vault_to_template(template) # not run

        return template # step 3
```

and additionally this `if not template.startswith((PATH_TOKEN, VAULT_TOKEN)):` doesn't actually capture _which_ token is inside the string if it does succeed.
So you still have to check the startswith again (up to 2 more times for a total of 4), which makes the code less efficient sure but more importantly (and this is obviously subjective) it doesn't feel right to me, like "code smell" or whatever its called. 

The only thing we'd be able to do by adding the 'bouncer pattern' is share the line 
`reference, _, conversion = vault_reference.partition(CONVERSION_TOKEN)` 
which debatably makes it more readable because one of the two paths doesn't even need `reference`, and would prefer that it was written like: `_, _, conversion = template.partition(CONVERSION_TOKEN)`.

To me the code reads like a simple 2 step switch case with a default, or a pattern match with a `_ => return template`

Reply:
Ah. I see where there's some confusion. 

Those two samples of code are actually **not** the same number of operations when step 3 is the final outcome, which will be the case a large percentage (if not all) of the time after we've fully migrated to Jinja and stopped recommending/using `$.` and `vault:` (even though the codebase would still support it for a while)

The former   --`if not template.startswith((PATH_TOKEN, VAULT_TOKEN))`-- is one `startswith` call that's passed a tuple input. The latter is two full `startswith` calls.  [I whipped up this gist quickly to illustrate the difference](https://gist.github.com/ubaniabalogun/2a952937ba27c24333252846e78f01dc) 

However, that isn't to say that the former is always absolutely faster than the latter. It depends on what our expected input is. Tweeking the `test_string` in the Gist will show our expected performance for `$.` & `vault:` (our currently most used use-cases) vs `{context.}` (our future most used use-cases). Which is why ultimately, optimization is tricky in this codebase and not a valid reason to duplicate the `partition(CONVERSION)` call in this function. 

"Code smell" is a better reason to consider the duplication of the partition logic, but yes also more subjective. I initially flagged this section because the duplication of the partition logic seemed like a "code smell" as it took a minute to figure out if those two calls were doing the same thing or not.